common/queue_test.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestNewQueue(t *testing.T) {
+	q := NewQueue[int]()
+
+	assert.NotNil(t, q, "NewQueue() should return a non-nil queue.")
+	assert.True(t, q.IsEmpty(), "A new queue should be empty.")
+	assert.Equal(t, 0, q.Len(), "A new queue should have a size of 0.")
+}
+
+func TestQueue_Push(t *testing.T) {
+	q := NewQueue[int]()
+
+	q.Push(4)
+	q.Push(5)
+
+	assert.Equal(t, 4, q.Peek())
+	assert.False(t, q.IsEmpty())
+}
+
+func TestQueue_Pop(t *testing.T) {
+	q := NewQueue[int]()
+	q.Push(4)
+	q.Push(5)
+	require.Equal(t, 4, q.Peek())
+	require.False(t, q.IsEmpty())
+
+	val := q.Pop()
+	assert.Equal(t, 4, val)
+	assert.Equal(t, 5, q.Peek())
+
+	val = q.Pop()
+	assert.Equal(t, 5, val)
+	assert.True(t, q.IsEmpty())
+}
+
+func TestQueue_PopNewQueue(t *testing.T) {
+	q := NewQueue[int]()
+	assert.True(t, q.IsEmpty(), "New queue should be empty.")
+
+	assert.Panics(t, func() {
+		q.Pop()
+	}, "Pop should panic when called on an empty queue.")
+}
+
+func TestQueue_PopEmptyQueue(t *testing.T) {
+	q := NewQueue[int]()
+	q.Push(3)
+	val := q.Pop()
+	assert.Equal(t, 3, val)
+	assert.True(t, q.IsEmpty())
+
+	assert.Panics(t, func() {
+		q.Pop()
+	}, "Pop should panic when called on an empty queue.")
+}
+
+func TestQueue_Peek(t *testing.T) {
+	q := NewQueue[int]()
+	q.Push(4)
+
+	val := q.Peek()
+	assert.Equal(t, 4, val)
+
+	// Peek again to ensure it doesn't remove the element.
```

Comment:
This comment is what this test is about I guess - the test name IMO should be TestQueuePeekIdempotent. I would recommend another test to verify that q.Peek() returns 4. Alternatively, don't test the idempotency and just end the test at `assert.Equal(t, 4, val)`